src/main/scala/inox/solvers/z3/Z3Unrolling.scala

```diff
@@ -68,7 +68,11 @@ trait Z3Unrolling extends AbstractUnrollingSolver { self =>
   private case class ModelWrapperImpl(model: Z3Model) extends super.ModelWrapper {
     private val ex = new underlying.ModelExtractor(model)
 
-    def extractConstructor(v: Z3AST, tpe: t.ADTType): Option[Identifier] = model.eval(v).flatMap {
+    private def tryZ3[T](res: => T): Option[T] =
+      // @nv: Z3 sometimes throws an exception when functions are called after Z3 has been canceled
+      try { Some(res) } catch { case e: Z3Exception if e.getMessage == "canceled" => None }
```

Comment:
Nit: should this rather be something like `tryEval`? It seems to be always used as a wrapper for `model.eval(...)`.

Reply:
Good suggestion!

Reply:
I see we now have a bunch of different implementations of `tryZ3` variants. Could we consolidate them all into a single implementation in `Z3Native.scala`? You can create a companion object if it makes more sense.

You can also provide a `tryZ3Opt` if you prefer that rather than calling `.flatten` at `tryZ3` call-sites.